cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -7,15 +7,33 @@ private import DataFlowDispatch
  * A data flow node that occurs as the argument of a call and is passed as-is
  * to the callable. Instance arguments (`this` pointer) are also included.
  */
-class ArgumentNode extends InstructionNode {
+class ArgumentNode extends Node {
   ArgumentNode() {
-    exists(CallInstruction call |
-      instr = call.getAnArgument()
-      or
-      instr.(ReadSideEffectInstruction).getPrimaryInstruction() = call
-    )
+    // To avoid making this class abstract, we enumerate its values here.
+    this instanceof PrimaryArgumentNode or
+    this instanceof SideEffectArgumentNode
```

Comment:
Why can't this class be abstract? It's in a file called `DataFlowPrivate.qll`, so hopefully it's not public.

Reply:
It's not a private class, though (as DataFlowImpl needs it). I figured that that the same argument for why we didn't make ParameterNode abstract (#3123 (comment)) would apply here as well.

Reply:
Done in 6c87b08c69d10ecebc2889c5d4c1c4ffa09518f3.